### PR TITLE
Persist FTP connection

### PR DESCRIPTION
**Pull request by necrophonic**
Changes behaviour to establish a single FTP connection and leave it open for subsequent messages. If the connection has not been opened or has closed it will attempt to re-establish it.

### CODE REVIEW

File: app/helpers/ftp.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        self.logger = logger
+        return
+
+    def get_connection(self):
+        """Connect checks whether an ftp connection is already open and, if
+        not, attempts to open a new one.
+        """
+        if self._conn is None:
+            # No connection at all
+            self.logger.info("Establishing new ftp connection", host=self.host)
```
Comment (eldeal) [inline app/helpers/ftp.py]: FTP should be capitalized in line with other log lines?

File: app/helpers/ftp.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    def _connect(self):
+        self._conn = FTP(self.host)
+        self._conn.login(user=self.user, passwd=self.passwd)
+        return self._conn
+
+    def deliver_binary(self, folder, filename, data):
+        """Delivery binary delivers a single binary file to the given folder
+        """
+        stream = io.BytesIO(data)
+        conn = self.get_connection()
```
Comment (eldeal) [inline app/helpers/ftp.py]: Is there really a need to get/check the connection for each binary or could it not be done in unzip_and_deliver before calling this method?

File: app/processors/common_software_processor.py [hunk 3]
```diff
@@ -33,12 +35,18 @@ def transform(self):
...
+        return self.ftp.unzip_and_deliver(folder, zip_contents)
 
     def process(self):
         zip_contents = self.transform()
         if zip_contents is None:
             return False
 
         return self.deliver_zip(zip_contents)
+
+    def get_ftp_folder(self, survey):
```
Comment (eldeal) [inline app/processors/common_software_processor.py]: Not actually sure if this is used by anything, but as it was already there I guess can leave for now, may need to investigate it's usefulness though.

**Comment by necrophonic** [comment]
Fixes #19

File: app/helpers/ftp.py (hunk 2 above)
Comment (necrophonic) [inline app/helpers/ftp.py]: I guess I was going at the slightly generic run of this being used in other libs (it'd need to be ported to the other downstreams) so some may want to deliver a single file and not use the zip method.

**Comment by necrophonic** [comment]
Will need to be ported to the other ``sdx-downstream-XX`` repos

File: app/processors/common_software_processor.py (hunk 3 above)
Comment (necrophonic) [inline app/processors/common_software_processor.py]: You can send a "heartbeat" survey type through I believe (not sure if they still do) and those drop to a separate folder (apparently)

File: app/helpers/ftp.py (hunk 2 above)
Comment (eldeal) [inline app/helpers/ftp.py]: Worth performance testing each way as we're doing this for performance benefit, but I take your point 👍

File: app/processors/common_software_processor.py (hunk 3 above)
Comment (eldeal) [inline app/processors/common_software_processor.py]: Yeah I don't think anyone actually does was my point. Might come back in to use as part of the synthetic transaction work upcoming, it's not doing any harm there.

**Comment by necrophonic** [comment]
**Note** Not to be merged until initial performance test is run

**Pull request merged**